docs/pallas/gpu/pipelining.md

```diff
@@ -0,0 +1,325 @@
+- `max_concurrent_steps` controls the maximum number of concurrent memory transfers. Using additional concurrent steps will consume more SMEM to hold temporary buffers, so this option should be used carefully.
+- `delay_release` allows the user to specify an additional number of iterations to wait before the buffer is re-used by the pipeline. For example, a buffer copied into SMEM on iteration 0 with `delay_release=1` and `max_concurrent_steps=2` will not be re-used until iteration 3, as opposed to iteration 2 for a standard double-buffered strategy. This is useful for certain optimizations such as allowing multiple async matmuls in flight to keep the tensor core pipeline filled.
+
+#### Compatibility API using `pl.pallas_call`
+
+As an alternative to `emit_pipeline` and to maintain compatibility with Pallas TPU, Mosaic GPU also implements the existing `pl.pallas_call` API for pipelining. Pipelining with `pl.pallas_call` directly requires the user to pass in a `plgpu.GPUCompilerParams` object as the `compiler_params` argument, which specifies the following options that are relevant for pipelining:
+- `dimension_semantics`: A tuple of `Literal['parallel', 'sequential']` that specifies iteration semantics for each grid dimension. `parallel` will partition the corresponding dimension over the CUDA grid, and `sequential` dimensions will be pipelined sequentially. **Note that if no dimensions are marked `sequential`, no pipelining will happen!**
+- `max_concurrent_steps`: the number of pipeline stages; identical to the option in `emit_pipeline`.
+- `delay_release`: the number of steps to wait before buffer re-use; idential to the option in `emit_pipeline`.
+
+We recommend that users use `plgpu.kernel` rather than `pl.pallas_call` as `plgpu.kernel` supports more features (such as specifying the number of warpgroups and warp specialization).
+
+<!-- #endregion -->
+
+<!-- #region id="Qp3X6wylJtoa" -->
+### GPU Memory Spaces
+
+Refs can exist in one of 3 memory spaces, specified by passing in the appropriate memory space into the `BlockSpec`, i.e. `BlockSpec(memory_space=plgpu.GPUMemorySpace.GMEM)`.
+
+- `plgpu.GPUMemorySpace.SMEM` allocates a Ref in Shared Memory (SMEM). SMEM Refs can be dereferenced using array indexing syntax to store values in registers for compute, i.e. `x = y_ref[...]`.
+
+- `plgpu.GPUMemorySpace.GMEM` allocates a Ref in Global Memory (GMEM/HBM). Any Refs allocated in GMEM are not pipelined, and values cannot be accessed directly using array indexing operations. Instead, GMEM must be accessed via SMEM using `plgpu.copy_gmem_to_smem` for reading, or `plgpu.copy_smem_to_gmem` for writing, or pipelined into SMEM using `plgpu.emit_pipeline`.
+
+- `plgpu.GPUMemorySpace.TMEM` allocates a Ref in Tensor Memory (TMEM). This is a TensorCore Gen 5 (Blackwell) feature for storing arguments and results of asynchronous matrix multiplication instructions. Like GMEM, TMEM cannot be accessed directly but must first be copied to/from SMEM.
```

Comment:
The part about accessing TMEM directly is false. It can be read from/written to registers and in fact it is the only way we support today

Reply:
Perhaps consider skipping the TMEM discussion here altogether since it has little to do with pipelining? It's a better fit for the reference doc

Reply:
Done - removing TMEM altogether.